proposals/p1964.md

```diff
@@ -0,0 +1,373 @@
+
+## Proposal
+
+Put character literals in single quotes, like `'a'`. Character literals work
+like numeric literals:
+
+-   Every different literal value has its own type.
+-   The bit width is determined by the type of the variable the literal is
+    assigned to, not the literal itself. Follows the plan from #1934.
+-   A character literal will model single Unicode code points that have a single
+    concrete numerical representation. We will not be supporting other
+    formulations like code unit sequences or grampheme clusters as these will be
+    modeled with normal string literals.
+
+## Details
+
+-   A character literal is a sequence enclosed with single quotes delimiter ('),
+    of UTF-8 code units that must be a valid encoding. This matches
+    [the UTF-8 encoding of Carbon source files](https://github.com/carbon-language/carbon-lang/blob/trunk/proposals/p0142.md#character-encoding).
+-   A character literal must encode exactly one code point.
+-   It supports addition and subtraction. These operations produce another code
+    point literal, or produce an error if the result is out of range.
+-   Character literals support some back-slash (`\`) escape sequences, including
+    `\t`, `\n`, `\r`, `\"`, `\'`, `\\`, `\0`, and `\u{HHHH...}`. See
+    [String Literals: Escape sequence](https://github.com/carbon-language/carbon-lang/blob/trunk/proposals/p0199.md#escape-sequences).
+
+We will not support:
+
+-   character literals that don't contain exactly one Unicode code point;
+-   multi-line literals;
+-   "raw" literals (using #'x'#);
+-   `\x` escape sequences;
+-   character literals with a single quote (`'`) or back-slash (`\`), except as
+    part of an escape sequence
+-   empty character literals (`''`);
+-   ASCII control codes (0...31), including whitespace characters other than
+    word space (tab, line feed, carriage return, form feed, and vertical tab),
+    except when specified with an escape sequence.
+
+### Types
+
+For the time being we will support character types `Char8`, `Char16`, and
+`Char32` that will hold both code units and code points, and will leave the
+different UTF-encoding code unit types to another proposal. See
+[UTF code unit types proposal](#utf-code-unit-types-proposal).
+
+We will have the type `CharN` and only support literals that map directly to the
+complete value of a code point.
+
+```
+let allowed: CharN = 'a';
+```
+
+The above is allowed because the type of `'a'` is the character literal
+consisting of the single Unicode code point 97, which can be converted to
+`CharN` since 97 is less than or equal to 0x7F.
+
+### Operations
+
+Character literals representing a single code point support the following
+operators:
+
+-   Comparison: `<`, `>`, `<=`, `>=` `==`
+-   Plus: `+`. This doesn't concatenate, but allows numerically adjusting the
+    value:
+    -   Only one operand may be a character literal, the other must be an
+        integer literal.
```

Comment:
One consequence appears to be that this is invalid:

```
fn Digit(n: i8) -> Char32 {
  return '0' + n;
}
```

... and something like `return ('0' as Char32) + n;` would be needed instead. I think I'm OK with that, but I expect it to be a minor source of friction as that kind of usage is fairly common in C++.

Reply:
Given that `'0' + n` has a value that is only known at runtime, what type should it be? Using the type of `n` here is a bit worrisome, due to overflow. I would be fine with saying the result would be `Char32`, but maybe that would only make sense for some types of `n`? For example if `n: i64`, a `Char32` result would be surprising.

Reply:
I think I like keeping this super-explicit for now (requiring a cast to a specific sized type). We can try to add better defaults if in practice this friction is something users dislike. I'm somewhat hopeful that instead we can have a really easily discovered (and targeted in migration) API for mapping to digits, and avoid how much this comes up in practice. But it seems easy to address if it does come up.